test/sql/test_cte.py

```diff
@@ -1377,3 +1378,333 @@ def test_no_alias_construct(self):
+        )
+
+    def test_nesting_cte_at_top_level(self):
+        nesting_cte = select([literal(1).label("val")]).cte(
+            "nesting_cte", nesting=True
+        )
+        cte = select([literal(2).label("val")]).cte("cte")
+        stmt = select([nesting_cte.c.val, cte.c.val])
+
+        self.assert_compile(
+            stmt,
+            "WITH nesting_cte AS (SELECT %(param_1)s AS val)"
+            ", cte AS (SELECT %(param_2)s AS val)"
+            " SELECT nesting_cte.val, cte.val AS val_1 FROM nesting_cte, cte",
+        )
+
+    def test_double_nesting_cte_in_cte(self):
+        """
+        Validate that the SELECT in the 2nd nesting CTE does not render
+        the 1st CTE.
+
+        It implies that nesting CTE level is taken in account.
+        """
+        select_1_cte = select([literal(1).label("inner_cte")]).cte(
+            "nesting_1", nesting=True
+        )
+        select_2_cte = select([literal(2).label("inner_cte")]).cte(
+            "nesting_2", nesting=True
+        )
+
+        stmt = select(
+            [
+                select(
+                    [
+                        select_1_cte.c.inner_cte.label("outer_1"),
+                        select_2_cte.c.inner_cte.label("outer_2"),
+                    ]
+                ).cte("cte")
+            ]
+        )
+
+        self.assert_compile(
+            stmt,
+            "WITH cte AS ("
+            "WITH nesting_1 AS (SELECT %(param_1)s AS inner_cte)"
+            ", nesting_2 AS (SELECT %(param_2)s AS inner_cte)"
+            " SELECT nesting_1.inner_cte AS outer_1"
+            ", nesting_2.inner_cte AS outer_2"
+            " FROM nesting_1, nesting_2"
+            ") SELECT cte.outer_1, cte.outer_2 FROM cte",
+        )
+
+    def test_double_nesting_cte_with_cross_reference_in_cte(self):
+        select_1_cte = select([literal(1).label("inner_cte_1")]).cte(
+            "nesting_1", nesting=True
+        )
+        select_2_cte = select(
+            [(select_1_cte.c.inner_cte_1 + 1).label("inner_cte_2")]
+        ).cte("nesting_2", nesting=True)
+
+        # 1 next 2
+
+        nesting_cte_1_2 = select([select_1_cte, select_2_cte]).cte("cte")
+        stmt_1_2 = select([nesting_cte_1_2])
+        self.assert_compile(
+            stmt_1_2,
+            "WITH cte AS ("
+            "WITH nesting_1 AS (SELECT %(param_1)s AS inner_cte_1)"
+            ", nesting_2 AS (SELECT nesting_1.inner_cte_1 + %(inner_cte_1_1)s"
+            " AS inner_cte_2 FROM nesting_1)"
+            " SELECT nesting_1.inner_cte_1 AS inner_cte_1"
+            ", nesting_2.inner_cte_2 AS inner_cte_2"
+            " FROM nesting_1, nesting_2"
+            ") SELECT cte.inner_cte_1, cte.inner_cte_2 FROM cte",
+        )
+
+        # 2 next 1
+
+        nesting_cte_2_1 = select([select_2_cte, select_1_cte]).cte("cte")
+        stmt_2_1 = select([nesting_cte_2_1])
+        self.assert_compile(
+            stmt_2_1,
+            # FIXME: nesting_1 is generated 2 times
```

Comment:
Expected
```sql
WITH cte AS (
	WITH nesting_1 AS (
		SELECT
			%(param_1) s AS inner_cte_1
	)
	, nesting_2 AS (
		SELECT
			nesting_1.inner_cte_1 + %(inner_cte_1_1) s AS inner_cte_2
		FROM
			nesting_1
	)
	SELECT
		nesting_2.inner_cte_2 AS inner_cte_2
		, nesting_1.inner_cte_1 AS inner_cte_1
	FROM
		nesting_2
		, nesting_1
)
SELECT
	cte.inner_cte_2
	, cte.inner_cte_1
FROM
	cte
```
Actual:
```sql
WITH cte AS (
	WITH nesting_2 AS (
		WITH nesting_1 AS (
			SELECT
				%(param_1) s AS inner_cte_1
		)
		SELECT
			nesting_1.inner_cte_1 + %(inner_cte_1_1) s AS inner_cte_2
		FROM
			nesting_1
	)
	, nesting_1 AS (
		SELECT
			%(param_1) s AS inner_cte_1
	)
	SELECT
		nesting_2.inner_cte_2 AS inner_cte_2
		, nesting_1.inner_cte_1 AS inner_cte_1
	FROM
		nesting_2
		, nesting_1
)
SELECT
	cte.inner_cte_2
	, cte.inner_cte_1
FROM
	cte
```

Reply:
I change the approach, I let this behavior possible as it is syntactically possible so it should not be impossible. And it is hard to auto fix without deep changes for no real benefits. I prefer to let the developer the ownership of which order matter. 

As shown by the first assertion, changing the column order is a first workaround. Another one shown by the second assertion uses `add_cte`. A developer has the possibility to refer a CTE before the query does and thus define the order.